File: backend/app/services/xai_explainer.py
"""
Natural-language explainability layer for the transparent priority scoring model.

Generates human-readable audit trails from SHAP-style factor breakdowns — no
external LLM required. Officers can review exactly why each incident received
its priority classification.
"""
from typing import Any, Dict, List


def _needs_summary(needs: Dict[str, bool]) -> List[str]:
    labels = {
        "food": "food supplies",
        "water": "clean water",
        "medical": "medical kits",
        "shelter": "emergency shelter",
        "rescue_team": "rescue teams",
    }
    return [labels[k] for k, v in needs.items() if v]


def explain_incident(incident: Dict[str, Any]) -> Dict[str, Any]:
    """Build a structured natural-language explanation for one incident."""
    score = incident.get("priority_score", 0)
    level = incident.get("priority_level", "medium")
    disaster = str(incident.get("disaster_type", "other")).replace("_", " ")
    name = incident.get("name", "Unknown zone")
    affected = incident.get("affected_count", 0)
    injured = incident.get("injured_count", 0)
    children = incident.get("children_count", 0)
    elderly = incident.get("elderly_count", 0)
    status = str(incident.get("status", "submitted")).replace("_", " ")
    factors: List[Dict] = incident.get("shap_breakdown") or []
    anomalies: List[str] = incident.get("anomaly_flags") or []
    needs = incident.get("needs") or {}

    increases = [f for f in factors if f.get("direction") == "increase"]
    decreases = [f for f in factors if f.get("direction") == "decrease"]
    top_increase = increases[0] if increases else None

    summary = (
        f"{name} is classified as {level} priority ({score}/100) due to a "
        f"{disaster} event affecting {affected:,} people, including "
        f"{injured} reported injuries and {children + elderly} vulnerable "
        f"individuals (children/elderly)."
    )

    reasoning_steps: List[str] = []
    for i, f in enumerate(factors, 1):
        sign = "+" if f.get("direction") == "increase" else ""
        pct = f.get("contribution_pct", 0)
        reasoning_steps.append(
            f"Step {i}: {f.get('label')} → {sign}{pct} points "
            f"({'raises' if f.get('direction') == 'increase' else 'lowers'} urgency)."
        )

    if top_increase:
        primary = (
            f"The dominant driver is \"{top_increase.get('label')}\" "
            f"(+{top_increase.get('contribution_pct')} pts), which heavily "
            f"influenced the {level} classification."
        )
    else:
        primary = "No single factor dominated; score reflects combined baseline severity."

    requested = _needs_summary(needs)
    if requested:
        needs_line = f"Requested resources: {', '.join(requested)}."
    else:
        needs_line = "No specific resource requests were flagged."

    if level in ("critical", "high"):
        dispatch = (
            f"Recommend immediate dispatch — escalate to active response. "
            f"Current status: {status}. Prioritize nearest available units."
        )
    elif level == "medium":
        dispatch = (
            f"Schedule staged response within standard SLA. "
            f"Monitor for escalation triggers. Status: {status}."
        )
    else:
        dispatch = (
            f"Low urgency — maintain monitoring and pre-position resources if deficit persists. "
            f"Status: {status}."
        )

    risk_factors = [f["label"] for f in increases if f.get("contribution_pct", 0) >= 10]
    mitigating_factors = [f["label"] for f in decreases]

    confidence = 92.0
    if anomalies:
        confidence -= min(25, len(anomalies) * 8)
    confidence = max(55.0, round(confidence, 1))

    return {
        "summary": summary,
        "primary_driver": primary,
        "reasoning_steps": reasoning_steps,
        "dispatch_recommendation": dispatch,
        "needs_summary": needs_line,
        "risk_factors": risk_factors,
        "mitigating_factors": mitigating_factors,
        "confidence_pct": confidence,
        "anomaly_warnings": anomalies,
    }


def aggregate_feature_importance(incidents: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """Average absolute SHAP contribution per feature label across all incidents."""
    totals: Dict[str, List[float]] = {}
    for inc in incidents:
        for f in inc.get("shap_breakdown") or []:
            label = f.get("label", "Unknown")
            totals.setdefault(label, []).append(abs(f.get("contribution_pct", 0)))

    aggregated = [
        {
            "label": label,
            "avg_contribution": round(sum(vals) / len(vals), 1),
            "incident_count": len(vals),
        }
        for label, vals in totals.items()
    ]
    aggregated.sort(key=lambda x: x["avg_contribution"], reverse=True)
    return aggregated

File: frontend/src/utils/xaiExplainer.ts
import type { Incident, NeedsRequested } from "../types";
import type { XAIExplanation } from "../types/xai";

function needsSummary(needs: NeedsRequested): string[] {
  const labels: Record<keyof NeedsRequested, string> = {
    food: "food supplies",
    water: "clean water",
    medical: "medical kits",
    shelter: "emergency shelter",
    rescue_team: "rescue teams",
  };
  return (Object.keys(labels) as (keyof NeedsRequested)[])
    .filter((k) => needs[k])
    .map((k) => labels[k]);
}

/** Client-side mirror of backend xai_explainer — works offline with mock data */
export function explainIncident(incident: Incident): XAIExplanation {
  const { priority_score: score, priority_level: level, disaster_type: disaster, name } = incident;
  const { affected_count: affected, injured_count: injured, children_count: children, elderly_count: elderly } = incident;
  const status = incident.status.replace(/_/g, " ");
  const factors = incident.shap_breakdown ?? [];
  const anomalies = incident.anomaly_flags ?? [];

  const increases = factors.filter((f) => f.direction === "increase");
  const decreases = factors.filter((f) => f.direction === "decrease");
  const topIncrease = increases[0];

  const summary =
    `${name} is classified as ${level} priority (${score}/100) due to a ` +
    `${disaster.replace(/_/g, " ")} event affecting ${affected.toLocaleString()} people, including ` +
    `${injured} reported injuries and ${children + elderly} vulnerable individuals (children/elderly).`;

  const reasoning_steps = factors.map((f, i) => {
    const sign = f.direction === "increase" ? "+" : "";
    const verb = f.direction === "increase" ? "raises" : "lowers";
    return `Step ${i + 1}: ${f.label} → ${sign}${f.contribution_pct} points (${verb} urgency).`;
  });

  const primary_driver = topIncrease
    ? `The dominant driver is "${topIncrease.label}" (+${topIncrease.contribution_pct} pts), which heavily influenced the ${level} classification.`
    : "No single factor dominated; score reflects combined baseline severity.";

  const requested = needsSummary(incident.needs);
  const needs_summary = requested.length
    ? `Requested resources: ${requested.join(", ")}.`
    : "No specific resource requests were flagged.";

  let dispatch_recommendation: string;
  if (level === "critical" || level === "high") {
    dispatch_recommendation =
      `Recommend immediate dispatch — escalate to active response. Current status: ${status}. Prioritize nearest available units.`;
  } else if (level === "medium") {
    dispatch_recommendation =
      `Schedule staged response within standard SLA. Monitor for escalation triggers. Status: ${status}.`;
  } else {
    dispatch_recommendation =
      `Low urgency — maintain monitoring and pre-position resources if deficit persists. Status: ${status}.`;
  }

  const risk_factors = increases.filter((f) => f.contribution_pct >= 10).map((f) => f.label);
  const mitigating_factors = decreases.map((f) => f.label);

  let confidence_pct = 92;
  if (anomalies.length) confidence_pct -= Math.min(25, anomalies.length * 8);
  confidence_pct = Math.max(55, Math.round(confidence_pct * 10) / 10);

  return {
    summary,
    primary_driver,
    reasoning_steps,
    dispatch_recommendation,
    needs_summary,
    risk_factors,
    mitigating_factors,
    confidence_pct,
    anomaly_warnings: anomalies,
  };
}

export function aggregateFeatureImportance(incidents: Incident[]) {
  const totals = new Map<string, number[]>();
  for (const inc of incidents) {
    for (const f of inc.shap_breakdown ?? []) {
      const list = totals.get(f.label) ?? [];
      list.push(Math.abs(f.contribution_pct));
      totals.set(f.label, list);
    }
  }
  return [...totals.entries()]
    .map(([label, vals]) => ({
      label,
      avg_contribution: Math.round((vals.reduce((a, b) => a + b, 0) / vals.length) * 10) / 10,
      incident_count: vals.length,
    }))
    .sort((a, b) => b.avg_contribution - a.avg_contribution);
}

export function buildXAIAnalytics(incidents: Incident[]) {
  const explained = incidents.map((incident) => ({
    ...incident,
    explanation: explainIncident(incident),
  }));
  const scores = explained.map((i) => i.priority_score);
  return {
    model_version: "1.0.0",
    model_type: "Transparent Weighted Feature Model (SHAP-compatible)",
    methodology:
      "Additive scoring from disaster base weight, population impact, injuries, vulnerable demographics, and resource deficit. Each factor's signed contribution is exposed for audit.",
    incidents: explained,
    aggregate_features: aggregateFeatureImportance(incidents),
    stats: {
      total: explained.length,
      critical: explained.filter((i) => i.priority_level === "critical").length,
      avg_score: scores.length ? Math.round((scores.reduce((a, b) => a + b, 0) / scores.length) * 10) / 10 : 0,
      anomalies: explained.filter((i) => i.anomaly_flags.length > 0).length,
    },
  };
}
